### PR TITLE
pageserver: use PITR GC cutoffs as authoritative

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Pageserver GC uses a size-based condition (GC "horizon" in addition to time-based "PITR").

Eventually we plan to retire the size-based condition: https://github.com/neondatabase/neon/issues/6374

Currently, we always apply the more conservative of the two, meaning that tenants always retain at least 64MB of history (default horizon), even after a very long time has passed.  This is particularly acute in cases where someone has dropped tables/databases, and then leaves a database idle: the horizon can prevent GCing very large quantities of historical data (we already account for this in synthetic size by ignoring gc horizon).

We're not entirely removing GC horizon right now because we don't want to 100% rely on standby_horizon for robustness of physical replication, but we can tweak our logic to avoid retaining that 64MB LSN length indefinitely.

## Summary of changes

- Rework `Timeline::find_gc_cutoffs`, with new logic:
  - If there is no PITR set, then use `DEFAULT_PITR_INTERVAL` (1 week) to calculate a time threshold.  Retain either the horizon or up to that thresholds, whichever requires less data.
  - When there is a PITR set, and we have unambiguously resolved the timestamp to an LSN, then ignore the GC horizon entirely.  For typical PITRs (1 day, 1 week), this will still easily retain enough data to avoid stressing read only replicas.

The key property we end up with, whether a PITR is set or not, is that after enough time has passed, our GC cutoff on an idle timeline will catch up with the last_record_lsn.

Using `DEFAULT_PITR_INTERVAL` is a bit of an arbitrary hack, but this feels like it isn't really worth the noise of exposing in TenantConfig.  We could just make it a different named constant though.  The end-end state will be that there is no gc_horizon at all, and that tenants with pitr_interval=0 would truly retain no history, so this constant would go away.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by koivunej — APPROVED** [review]

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -4974,58 +4975,71 @@ impl Timeline {
...
-                        // have no commit timestamp data available
-                        *self.get_latest_gc_cutoff_lsn()
-                    }
-                    LsnForTimestamp::NoData(lsn) => {
-                        debug!("nodata({})", lsn);
-                        // conservative, safe default is to remove nothing, when we
-                        // have no commit timestamp data available
-                        *self.get_latest_gc_cutoff_lsn()
-                    }
+            // If PITR is so large or `now` is so small that this overflows, we will retain no history (highly unexpected case)
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:4991]: overflow == underflow here right?

Comment (jcsp) [reply pageserver/src/tenant/timeline.rs]: Yes, in my head it was a negative overflow but you're right :-)  https://github.com/neondatabase/neon/pull/8365/commits/2c38053f13949578403d601ffd259bd07b705ef9

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -4974,58 +4975,81 @@ impl Timeline {
...
+                return Ok(GcCutoffs {
+                    pitr: self.get_last_record_lsn(),
+                    horizon: cutoff_horizon,
+                });
+            }
+        }
+
+        // Calculate a time-based limit on how much to retain:
+        // - if PITR interval is configured, then this is our cutoff.
+        // - if PITR interval is not configured, then we do a lookup
```
Comment (hlinnaka) [inline pageserver/src/tenant/timeline.rs:4990]: Does "not configured" mean `Duration::ZERO`? That sounds confusing. The default, if you create a new project and don't configure anything is something different, right?

**Review by hlinnaka — COMMENTED** [review]
Does the function comment need updating? 

Not this PR's fault, but the terminology with the time- and size-based cutoffs are really confusing. "pitr horizon", "cutoff", "retention period" all sound like synonyms to me. Some thoughts:

- How about renaming renaming the `cutoff_horizon` and `pitr` variables to `wal_size_cutoff` and `time_cutoff` or something similar?
- Does this function (`find_gc_cutoffs') really need to return both cutoff values? When perfoming GC, you really only need one Lsn cutoff. `Timeline::gc_timeline` keeps statistics of which cutoff value is used per layer, but I don't think that's a very interesting statistic.

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline.rs]: Agree that `configured` is ambiguous here.  I've changed this to "set" and cleaned up the function docstring.  f6ad06248

I think we can in future remove the default PITR in the pageserver: this policy really belongs to the control plane, and a sane default for the pageserver would be no PITR at all.

**Comment by jcsp** [comment]
>How about renaming renaming the cutoff_horizon and pitr variables to wal_size_cutoff and time_cutoff or something similar?
>Does this function (find_gc_cutoffs') really need to return both cutoff values? When perfoming GC, you really only need one Lsn cutoff. Timeline::gc_timeline` keeps statistics of which cutoff value is used per layer, but I don't think that's a very interesting statistic.

Yes to both, let's do this in a separate PR so that we have separate PRs for the functional change and the refactor.

**Pull request merged**